lld/MachO/BPSectionOrderer.h

```diff
@@ -14,13 +14,149 @@
+    return isec && !isec->data.empty() && isec->data.data();
+  }
+
+  llvm::ArrayRef<uint8_t> getSectionData() const override { return isec->data; }
+
+  llvm::ArrayRef<BPSymbol *> getSymbols() const override {
+    // Lazy initialization of symbol cache
+    if (symbolCache.empty()) {
+      for (const auto *sym : isec->symbols)
+        symbolCache.push_back(std::make_unique<BPSymbolMacho>(sym));
+    }
+    static std::vector<BPSymbol *> result;
+    result.clear();
+    for (const auto &sym : symbolCache)
+      result.push_back(sym.get());
+    return result;
+  }
+
+  void getSectionHash(llvm::SmallVectorImpl<uint64_t> &hashes,
+                      const llvm::DenseMap<const BPSectionBase *, uint64_t>
+                          &sectionToIdx) const override {
+    constexpr unsigned windowSize = 4;
+
+    // Convert BPSectionBase map to InputSection map
+    llvm::DenseMap<const InputSection *, uint64_t> machoSectionToIdx;
+    for (const auto &[sec, idx] : sectionToIdx)
+      if (auto *machoSec = llvm::dyn_cast<BPSectionMacho>(sec))
+        machoSectionToIdx[machoSec->getInputSection()] = idx;
+
+    // Calculate content hashes
+    size_t dataSize = isec->data.size();
+    for (size_t i = 0; i < dataSize; i++) {
+      auto window = isec->data.drop_front(i).take_front(windowSize);
+      hashes.push_back(xxHash64(window));
+    }
+
+    // Calculate relocation hashes
+    for (const auto &r : isec->relocs) {
+      if (r.length == 0 || r.referent.isNull() || r.offset >= isec->data.size())
+        continue;
+
+      uint64_t relocHash = getRelocHash(r, machoSectionToIdx);
+      uint32_t start = (r.offset < windowSize) ? 0 : r.offset - windowSize + 1;
+      for (uint32_t i = start; i < r.offset + r.length; i++) {
+        auto window = isec->data.drop_front(i).take_front(windowSize);
+        hashes.push_back(xxHash64(window) + relocHash);
+      }
+    }
+
+    llvm::sort(hashes);
+    hashes.erase(std::unique(hashes.begin(), hashes.end()), hashes.end());
+  }
+
+  const InputSection *getInputSection() const { return isec; }
+
+  static bool classof(const BPSectionBase *s) { return true; }
+
+private:
+  static uint64_t getRelocHash(
+      const Reloc &reloc,
+      const llvm::DenseMap<const InputSection *, uint64_t> &sectionToIdx) {
+    auto *isec = reloc.getReferentInputSection();
+    std::optional<uint64_t> sectionIdx;
+    auto sectionIdxIt = sectionToIdx.find(isec);
+    if (sectionIdxIt != sectionToIdx.end())
+      sectionIdx = sectionIdxIt->getSecond();
+
+    std::string kind;
+    if (isec)
+      kind = ("Section " + Twine(isec->kind())).str();
+
+    if (auto *sym = reloc.referent.dyn_cast<Symbol *>()) {
+      kind += (" Symbol " + Twine(sym->kind())).str();
+      if (auto *d = llvm::dyn_cast<Defined>(sym)) {
+        if (llvm::isa_and_nonnull<CStringInputSection>(isec))
+          return BPSectionBase::getRelocHash(kind, 0, isec->getOffset(d->value),
+                                             reloc.addend);
```

Comment:
We no longer check if this section is a cstring section. https://github.com/llvm/llvm-project/commit/6827a00d4dfe1fa7d479337b6192602744055686